# 01. Importing the libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 02. Importing the dataset

In [2]:
# create the path
path = r'C:\Users\Lumex\Documents\Achievement_4_Project\InstaCart Basket Analysis'
#import products dataset
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03. Data Consistency Checks

In [3]:
#Creating a mix data type column in a test dataframe
df_test = pd.DataFrame()

In [4]:
# creating a mixed datatype
df_test['mix'] = ['a', 'b', 1, True]

In [5]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [6]:
# The function for checking whether a dataframe contains any mixed-type columns
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
       print (col)

mix


C:\Users\Lumex\AppData\Local\Temp\ipykernel_10088\1216472489.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [7]:
#Convert the mixed-type  data to string.
df_test['mix'] = df_test['mix'].astype('str')

## 04. Missing value 

In [8]:
# Checking for missing values in the df_prods dataset
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [9]:
#Creating a subset of the dataframe containing the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True ]

In [10]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [11]:
#Running exploratory analysis on the df_prods dataframe
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [12]:
# finding the median for df_prods 
df_prods['prices'].median()

7.1

In [13]:
#checking for outliers in the df_prods dataframe.
df_prods[df_prods['prices'] > df_prods['prices'].mean() + (2 * df_prods['prices'].std())]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

In [14]:
# replacing the outliers with the median 7.1
df_prods = df_prods.replace({"prices":{14900.0:7.1, 99999.0:7.1}})

In [15]:
#verify if the missing values have been imputated
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      7.682245
std    14343.717401     38.316774       5.850282      4.200159
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000     25.000000

In [16]:
#Current number of rows & columns in df_prods dataframe
df_prods.shape

(49693, 5)

In [17]:
#Create a new dataframe as a subset of df_prods dataframe where the missing values in the product_name variable are excluded
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [18]:
df_prods_clean.shape

(49677, 5)

#### Observation about df_prods dataframe.
##### After running an exploratory analysis on the df_prods, some outliners were discovered in the data. The mean of the price column is 9.99 with a std of 453.52 and the outliers were replaced with the median. After the mean was replaced with the median, the mean of the column price was reduced to 7.6. There were 16 missing values in the df_prods dataset. There was mixed-type data also discovered in the datatype and it was converted to a string. Duplicated data was also removed to make the dataframe more consistent.

#### Duplicate values

In [19]:
#creates a new subset containing only the rows that are duplicated.
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [21]:
# the number of rows in the df_prods_clean datafrane
df_prods_clean.shape

(49677, 5)

In [22]:
#Create a new datafrme where there is no duplicate values
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [23]:
df_prods_clean_no_dups.shape

(49672, 5)

## Task 4.5

### Orders DataFrame

### Data Consistency Checks

In [24]:
# Run an exploratory analysis in the df_ords dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

#### Step 2: Observation about the exploratory analysis for df_ords
##### The data does not look odd and there is no outliers in the analysis.

In [25]:
#Checking the dataframe for mixed data type within a column/variable in the df_ords.
#Check for mixed-type data in your df_ords dataframe
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [26]:
df_ords.tail()

Unnamed: 0  order_id  user_id eval_set  order_number  \
3421078     3421078   2266710   206209    prior            10   
3421079     3421079   1854736   206209    prior            11   
3421080     3421080    626363   206209    prior            12   
3421081     3421081   2977660   206209    prior            13   
3421082     3421082    272231   206209    train            14   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
3421078                   5                 18                    29.0  
3421079                   4                 10                    30.0  
3421080                   1                 12                    18.0  
3421081                   1                 12                     7.0  
3421082                   6                 14                    30.0

#### Step 3: After looking through the orders dataframe, there is no mixed datatype found in the columns.

In [27]:
# Running a function to search for a mixed data type in a column in the df_ords.
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
       print (col)      

C:\Users\Lumex\AppData\Local\Temp\ipykernel_10088\3559709897.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Lumex\AppData\Local\Temp\ipykernel_10088\3559709897.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Lumex\AppData\Local\Temp\ipykernel_10088\3559709897.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Lumex\AppData\Local\Temp\ipykernel_10088\3559709897.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Lumex\AppData\Local

In [28]:
#checking the datatypes in the columns of df_ords
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 8 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   eval_set                object 
 4   order_number            int64  
 5   orders_day_of_week      int64  
 6   order_hour_of_day       int64  
 7   days_since_prior_order  float64
dtypes: float64(1), int64(6), object(1)
memory usage: 208.8+ MB


#### Step 4: Mixed-Typed Data
##### A function was to check if there there was mixed-type data in the df_ords. There was no mixed-typed data found in the dataframe.
##### The info() function also confirm that there were no mixed-type data in the dataframe.

In [29]:
# Step 5:Run a check for missing values in your df_ords dataframe.
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [30]:
df_ords

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
1                 1   2398795        1    prior             2   
2                 2    473747        1    prior             3   
3                 3   2254736        1    prior             4   
4                 4    431534        1    prior             5   
...             ...       ...      ...      ...           ...   
3421078     3421078   2266710   206209    prior            10   
3421079     3421079   1854736   206209    prior            11   
3421080     3421080    626363   206209    prior            12   
3421081     3421081   2977660   206209    prior            13   
3421082     3421082    272231   206209    train            14   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
1                         3                  7                    15.0  
2                         3                 12                    21.0  
3                         4                  7                    29.0  
4                         4                 15                    28.0  
...                     ...                ...                     ...  
3421078                   5                 18                    29.0  
3421079                   4                 10                    30.0  
3421080                   1                 12                    18.0  
3421081                   1                 12                     7.0  
3421082                   6                 14                    30.0  

[3421083 rows x 8 columns]

##### Step 5: There are 206,209 values missing in the days_since_prior_order rows out of the total of 3,421083 rows of the df_ords

In [31]:
#investigating the missing values based on user_id count summarized 
df_ords[df_ords['days_since_prior_order'].isnull()].groupby('user_id').count().describe()

Unnamed: 0  order_id  eval_set  order_number  orders_day_of_week  \
count    206209.0  206209.0  206209.0      206209.0            206209.0   
mean          1.0       1.0       1.0           1.0                 1.0   
std           0.0       0.0       0.0           0.0                 0.0   
min           1.0       1.0       1.0           1.0                 1.0   
25%           1.0       1.0       1.0           1.0                 1.0   
50%           1.0       1.0       1.0           1.0                 1.0   
75%           1.0       1.0       1.0           1.0                 1.0   
max           1.0       1.0       1.0           1.0                 1.0   

       order_hour_of_day  days_since_prior_order  
count           206209.0                206209.0  
mean                 1.0                     0.0  
std                  0.0                     0.0  
min                  1.0                     0.0  
25%                  1.0                     0.0  
50%                  1.0                     0.0  
75%                  1.0                     0.0  
max                  1.0                     0.0

#### Step 6: Address the missing values using an appropriate method.
##### The df_ords datafraame contains 3,421083, 7 columns and there are 206,209 values missing in the days_since_prior_order column, so i investigated the missing value by counting user_id which result to 206,209 count. The summarization suggest that there is one missing value in each unique user_id. There is no reason to remove the missing value because every unique user_id have only one missing value in the days_since_prior_order and the missing value is from the first order of each unique user_id.

In [32]:
# check for duplicate values in your df_ords data
df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

#### Step 7 and 8: There were no duplicate values found in df_ords after running the duplicate function on the dataframe.

In [33]:
# Remove unnamed column from df_ords_cleaned because it is not useful and it is the same with the index number.

df_ords_clean = df_ords.drop('Unnamed: 0', axis=1)

In [34]:
df_ords_clean

order_id  user_id eval_set  order_number  orders_day_of_week  \
0         2539329        1    prior             1                   2   
1         2398795        1    prior             2                   3   
2          473747        1    prior             3                   3   
3         2254736        1    prior             4                   4   
4          431534        1    prior             5                   4   
...           ...      ...      ...           ...                 ...   
3421078   2266710   206209    prior            10                   5   
3421079   1854736   206209    prior            11                   4   
3421080    626363   206209    prior            12                   1   
3421081   2977660   206209    prior            13                   1   
3421082    272231   206209    train            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

#### Step 9: Export the final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder.

In [35]:
#Export the cleaned datafrme for df_prods
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [36]:
#Export the cleaned datafrme for df_ords
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))